iree/turbine/kernel/wave/decompose_scan_ops.py

```diff
@@ -28,61 +32,148 @@ def get_graph_node(custom: CustomOp, graph: fx.Graph) -> fx.Node:
     return custom.fx_node
 
 
+def emit_local_inclusive_scan(
+    binary_fn: Callable,
+    scan_src: fx.Node,
+    graph: fx.Graph,
+    elements_per_thread: int,
+) -> list[fx.Node]:
+    """
+    Perform local inclusive scan for `n` elements per thread.
+    """
+    values = [
+        get_graph_node(Extract(scan_src, [i]), graph)
+        for i in range(elements_per_thread)
+    ]
+
+    for i in range(1, elements_per_thread):
+        values[i] = get_graph_node(binary_fn(values[i], values[i - 1]), graph)
+
+    return values
+
+
 def emit_global_scan(
     binary_fn: Callable,  # Supports only Add for now.
     src: fx.Node,
+    local_scan: list[fx.Node],
     graph: fx.Graph,
     subgroup_size: int,
     hardware_constraint: HardwareConstraint,
+    local_scan_size: int,
+    scan_dim: IndexSymbol,
 ) -> fx.Node:
     """
     Emit an intra-warp inclusive scan using butterfly pattern scan and masking.
+    In this algorithm, we are using two-pass approach.
+    If the number of elements per thread is one, then we just need one shuffle pass.
+    If there are more than one elements per thread, we need to calculate the exclusive
+        offset for the next lane elements using the information from the previous lane.
+        That is why we have another shuffle phase.
+
+    ToDo: Can we make it more efficient?
     """
-    init = src
+    lane_id = (
+        hardware_constraint.linearized_thread_id % hardware_constraint.threads_per_wave
+    )
+
+    scanop_result = last_local_scan_node = local_scan[-1]
```

Comment:
Seems like we are only using `last_local_scan_node` for it's index. If that is the case, any thoughts on refactoring to something like:
```
scanop_result = last_local_scan_node = local_scan[-1]
local_scan_index = get_custom(local_scan[-1]).index
```
I think it makes each variable clearer on what it is for and added bonus of not needing to do get_custom() + .index at every use